### PR TITLE
[homematic] Add Authentication

### LINKED ISSUE #8625: [homematic] CCU2 with Authentication
At least CCU2 firware 2.53.27 doesn't support authentication less access. There is an automatic login feature but this doesn't seem to work the same on the Remote API.

Please support authentication working with latest firmware.

### PR DESCRIPTION
**Pull request by EvilPingu**
# Description

This PR enable authorization (resolves #8625 ) when accessing Homematic Server (like CCU). I've added three new binding configuration values to enable this. Now it's possible to disable unauthorized accesses to the CCU. I added some documentation which is relevant for testing.

# Testing

I tested the new feature with raspberrymatic and OH running in a K8s environment. I thinks it's necessary to test it with other settings.

### CODE REVIEW

File: bundles/org.openhab.binding.homematic/README.md [hunk 1]
```diff
@@ -14,13 +14,22 @@ When the option "Restricted access" is used, some ports have to be added to the
...
+8181;
 8701;
 9292;
 ```
 
 Also the IP of the device running openHAB has to be set to the list of "IP addresses for restricted access".
 
-Also under `Home page > Settings > Control panel` with the menu `Security` the option `Authentication` has to be disabled as the binding does not support the configuration of `username` and `password`for the XML-RPC API.
+Also under `Home page > Settings > Control panel` with the menu `Security` the option `Authentication` has to be disabled if the option 'useAuthentication' is not set. This option may to be enabled 
+if the option 'useAuthentication' is set and BIN-RPC is not used. In this case a user and a password must be created. This can be done under `Home page > Settings > Control panel` with the menu `User management`. The new user should have the following configuration:
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/README.md:25]: Minor but needed,  every sentence should go on its own line. Maybe also have a second look at the grammer.

File: bundles/org.openhab.binding.homematic/README.md [hunk 2]
```diff
@@ -14,13 +14,22 @@ When the option "Restricted access" is used, some ports have to be added to the
...
+
+- User name - button for login: No
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/README.md:27]:
```suggestion
- Username - button for login: No
```

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/AuthenticationHandler.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    public AuthenticationHandler(HomematicConfig config, HttpClient httpClient) {
+        this.config = config;
+        this.httpClient = httpClient;
+    }
+
+    /**
+     * Add or remove the basic auth credetials to the AuthenticationStore if needed.
+     */
+    public synchronized void updateAuthenticationInformation(final URI uri) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/AuthenticationHandler.java:42]:
```suggestion
    public synchronized void updateAuthenticationInformation(final URI uri) throws IllegalStateException {
```

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/i18n/homematic_de.properties [hunk 4]
```diff
@@ -49,6 +49,12 @@ thing-type.config.homematic.bridge.wiredPort.label = Wired Port
 thing-type.config.homematic.bridge.wiredPort.description = Die Portnummer des HS485-Dienstes
 thing-type.config.homematic.bridge.xmlCallbackPort.label = XML-RPC Callback Port
 thing-type.config.homematic.bridge.xmlCallbackPort.description = Callback-Port des XML-RPC-Servers. Wenn kein Wert angegeben ist, startet xmlCallbackPort mit 9125 und zählt hoch
+thing-type.config.homematic.bridge.password.label = Password für Zugriff auf das Gateway
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/i18n/homematic_de.properties:52]: Please remove this file as translations should be added by the use of crowdin.

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml [hunk 5]
```diff
@@ -127,6 +127,21 @@
 				<default>2048</default>
 				<advanced>true</advanced>
 			</parameter>
+			<parameter name="useAuthentication" type="boolean">
+				<label>Use authentication to access the gateway</label>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml:131]: Labels should be: <25 chars, max 2-3 words and properly capitalized.
(after this is fixed, don't forget top run the mvn i18n plugin)

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml [hunk 6]
```diff
@@ -127,6 +127,21 @@
...
+				<description>If set to true, username and passwort is required.</description>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml:132]:
```suggestion
				<description>If set to true, username and password is required.</description>
```

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml [hunk 7]
```diff
@@ -127,6 +127,21 @@
...
+				<advanced>true</advanced>
+				<default>false</default>
+			</parameter>
+			<parameter name="userName" type="text">
+				<label>User name for accessing the gateway</label>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml:137]: Labels should be: <25 chars, max 2-3 words and properly capitalized. suggestion: "Gateway Username"
(after this is fixed, don't forget top run the mvn i18n plugin)

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml [hunk 8]
```diff
@@ -127,6 +127,21 @@
...
+				<advanced>true</advanced>
+			</parameter>
+			<parameter name="password" type="text">
+				<label>Password for accessing the gateway</label>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml:141]: Labels should be: <25 chars, max 2-3 words and properly capitalized.
(after this is fixed, don't forget top run the mvn i18n plugin)

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/AuthenticationHandler.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        final AuthenticationStore authStore = httpClient.getAuthenticationStore();
+
+        Authentication findAuthentication = authStore.findAuthentication("Basic", uri, Authentication.ANY_REALM);
+
+        if (config.getUseAuthentication()) {
+            if (findAuthentication == null) {
+                if (config.getPassword() == null || config.getUserName() == null) {
+                    throw new IllegalStateException("Username or password missing");
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/AuthenticationHandler.java:50]: Have not yet looked up the chain to see if this is catched somewhere, but be sure that it leads to a thingstatus with configuration error.

**Review by lsiepel — CHANGES_REQUESTED** [review]
Left some comments, nothing big.  Thanks.,

File: bundles/org.openhab.binding.homematic/README.md (hunk 1 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/README.md]: Done

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/i18n/homematic_de.properties (hunk 4 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/i18n/homematic_de.properties]: Done

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml (hunk 5 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml]: Done

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml (hunk 7 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml]: Done

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml (hunk 8 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml]: Done

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml (hunk 6 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/thing/bridge.xml]: Done

File: bundles/org.openhab.binding.homematic/README.md (hunk 2 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/README.md:31]: I copied this from the website where the new user needs to be created. That's why I leave it as it is

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/AuthenticationHandler.java (hunk 9 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/AuthenticationHandler.java]: Changed to configuration exception; Set configuration error state to binding

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/AuthenticationHandler.java (hunk 3 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/AuthenticationHandler.java]: Move the exception to the constructor; Throw a configuration exception

**Comment by EvilPingu** [comment]
> Left some comments, nothing big. Thanks.,

Thanks for your remarks. Changed the error handling a little bit

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicBridgeHandler.java [hunk 10]
```diff
@@ -132,6 +133,10 @@ private void initializeInternal() {
                         ex.getMessage(), ex);
                 disposeInternal();
                 scheduleReinitialize();
+            } catch (ConfigurationException ex) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, ex.getMessage());
+                logger.debug("Configuration error while initialize binding: {}", ex.getMessage(), ex);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicBridgeHandler.java:138]:
```suggestion
```

As the thing state is allready set, what is the point of logging and writing the stacktrace, is that needed?

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/i18n/homematic.properties [hunk 11]
```diff
@@ -54,3 +60,8 @@ thing-type.config.homematic.bridge.xmlCallbackPort.description = Callback port o
 
 channel-type.homematic.DUTY_CYCLE_RATIO.label = Duty Cycle
 channel-type.homematic.DUTY_CYCLE_RATIO.description = Current duty cycle usage
+
+# thing types config
+
+thing-type.config.homematic.bridge.username.label = Username for accessing the gateway
+thing-type.config.homematic.bridge.username.description = Username for accessing the gateway. Is required if the "useAuthentication" is true
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/i18n/homematic.properties:67]: Did you run the i18n plugin or was it manually added? This keys allready exist around line 52.

**Review by lsiepel — CHANGES_REQUESTED** [review]
My last comments, would be nice if @gerrieg can also have a look as codeowner.

File: bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/i18n/homematic.properties (hunk 11 above)
Comment (EvilPingu) [reply bundles/org.openhab.binding.homematic/src/main/resources/OH-INF/i18n/homematic.properties:67]: I run i18n plugin - however i have a look on this

**Comment by EvilPingu** [comment]
> My last comments, would be nice if @gerrieg can also have a look as codeowner.

all done

**Review by lsiepel — APPROVED** [review]
LGTM

**Pull request merged**